thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/cube/data/node/CubeNode.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.thirdeye.cube.data.node;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import java.util.List;
+import org.apache.pinot.thirdeye.cube.data.dbrow.DimensionValues;
+import org.apache.pinot.thirdeye.cube.data.dbrow.Dimensions;
+
+
+public interface CubeNode<N extends CubeNode> {
+
+  /**
+   * Returns the level of this node in the cube; level 0 the topmost level.
+   *
+   * @return the level of this node in the cube.
+   */
+  int getLevel();
+
+  /**
+   * Resets all values (e.g., baseline and current value) of this node.
+   */
+  void resetValues();
+
+  /**
+   * Updates all values when the child node is extracted from this node.
+   *
+   * @param node the child node to be extracted.
+   */
+  void removeNodeValues(N node);
```

Comment:
Explained in the above comment.

Reply:
Can you double check? @cyenjung

Reply:
If we use `CubeNode` here, we need to cast it to the same type as the current node. Because removing a ratio cube node from an additive cube node doesn't make sense and vice versa.

To prevent from such mistake happens during runtime (we want to report such error during compile time), it's a better idea to use an explicit class `N`. However, I want to bound it to `BaseCubeNode` instead of any arbitrary class. Hence `N extends BaseCubeNode`.